### PR TITLE
Add support for reading TIFFs with PlanarConfiguration=2

### PR DESCRIPTION
**Pull request by wiredfool**
* Rebase of #4641
* Add support for reading TIFFs with PlanarConfiguration=2

### CODE REVIEW

**Comment by wiredfool** [comment]
I'm afraid I don't have the setup to debug the windows side of this at the moment.

**Comment by kkopachev** [comment]
I am not familiar with Windows specifics here, but here is some observations:
Current code failed only on x86. On few runs it died with 'Access violation' error and on Python 3.9 it throws [Col out of range, max 99](https://github.com/python-pillow/Pillow/pull/5178/checks?check_run_id=1643657090#step:27:515l) error. I thought it's something wrong in calculation of accessing tile width and enabled `TRACE` to see what's up - and x86 runs all successful, but now it failed on all x64 [in different test](https://github.com/kkopachev/Pillow/runs/1647707573?check_suite_focus=true#step:27:516).
Thought that might be helpful.

**Comment by wiredfool** [comment]
@nulano Do you have any insight into this?

File: src/libImaging/TiffDecode.c [hunk 1]
```diff
@@ -558,65 +589,89 @@ ImagingLibTiffDecode(
...
+                        if (isYCbCr) {
+                            current_line = tile_length - tile_y - 1;
+                        } else {
+                            current_line = tile_y;
+                        }
+
+                        shuffler((UINT8*) im->image[tile_y + y] + x * im->pixelsize,
+                                 state->buffer + current_line * row_byte_size,
+                                 current_tile_width
+                                 );
```
Comment (nulano) [inline src/libImaging/TiffDecode.c:637]: The error doesn't show up when compiling with in a debug build (`setup.py build_ext --dubug`), I had to change `setup.py` line 852 to `Extension("PIL._imaging", files, extra_compile_args=["/Z7"], extra_link_args=["/DEBUG"]),` and copy the PDB to the PIL directory.

I'm not yet sure what is going on, but compiling with debug information shows that `state->buffer` is NULL on this line even though it was valid on line 587 for the realloc call.

Comment (nulano) [reply src/libImaging/TiffDecode.c:637]: Solved it, see https://github.com/wiredfool/Pillow/pull/6

File: src/libImaging/TiffDecode.c [hunk 2]
```diff
@@ -405,6 +402,11 @@ ImagingLibTiffDecode(
     TIFF *tiff;
     uint16 photometric = 0;  // init to not PHOTOMETRIC_YCBCR
     int isYCbCr = 0;
+    UINT8 planarconfig = 0;
+    UINT8 planes = 1;
```
Comment (nulano) [inline src/libImaging/TiffDecode.c:406]: Is there any reason to use `UINT8` for `planes`? It looks to me like it would make more sense to use `int`.

Comment (kkopachev) [reply src/libImaging/TiffDecode.c:406]: Yeah, should be just `int` :facepalm:

File: src/libImaging/TiffDecode.c [hunk 3]
```diff
@@ -501,6 +503,35 @@ ImagingLibTiffDecode(
 
     TIFFGetField(tiff, TIFFTAG_PHOTOMETRIC, &photometric);
     isYCbCr = photometric == PHOTOMETRIC_YCBCR;
+    TIFFGetFieldDefaulted(tiff, TIFFTAG_PLANARCONFIG, &planarconfig);
+
+    // YCbCr data is read as RGB by libtiff and we don't need to worry about planar storage in that case
+    // if number of bands is 1, there is no difference with contig case
+    if (planarconfig == PLANARCONFIG_SEPARATE &&
+        im->bands > 1 &&
+        photometric != PHOTOMETRIC_YCBCR) {
```
Comment (kkopachev) [inline src/libImaging/TiffDecode.c:512]:
```suggestion
        !isYCbCr) {
```

Comment (wiredfool) [reply src/libImaging/TiffDecode.c]: Thanks @nulano and @kkopachev

**Review by nulano — COMMENTED** [review]
Same as `UINT8 planes` -> `int planes`.

**Comment by kkopachev** [comment]
@wiredfool I have some changes I did on top of this PR to fix BigEndian skipped tests and a bit of refactoring. Do you want me to create PR into this branch or wait until this merged and open into Pillow:master afterwards for separate review?
changes are here: https://github.com/wiredfool/Pillow/compare/4641_rebase...kkopachev:4641_rebase

**Comment by wiredfool** [comment]
Might as well pr to this branch

**Pull request closed without merging**